### ISSUE TITLE
Add a :skip status to pre_commit hooks

### ISSUE THREAD

**Issue by reggieb** [brigade/overcommit#359]
We want to use the same overcommit settings across a number of projects. We are using https://github.com/EnvironmentAgency/before_commit to copy a common set of configuration files into a project, and run the initial setup.

Most of the projects are rails apps, but a few are standalone gems. We want to include brakeman when the project is a rails app, but skip the brakeman tests when the project is a gem.

At the moment, the pre_commit hook below works for us:

```ruby
module Overcommit::Hook::PreCommit
  class Brakeman < Base
    def run
      return :pass unless File.exist?('bin/rails')
      result = execute(command + [applicable_files.join(',')])
      return :pass if result.success?

      [:fail, result.stdout]
    end
  end
end
```

However, the resulting output for gem projects suggests that brakeman tests have been run. It would be good if we could do this:

```ruby
module Overcommit::Hook::PreCommit
  class Brakeman < Base
    def run
      return :skip unless File.exist?('bin/rails')
      result = execute(command + [applicable_files.join(',')])
      return :pass if result.success?

      [:fail, result.stdout]
    end
  end
end
```

I think modifying `Overcommit::Printer#print_results` method as shown below would achieve the desired effect, but I'm not sure if there would be any other knock on effects. 

```ruby
    def print_result(hook, status, output)
      case status
      when :pass
        log.success 'OK' unless hook.quiet?
      when :skip
        log.warning 'SKIPPED'
      when :warn
        log.warning 'WARNING'
        print_report(output, :bold_warning)
      when :fail
        log.error 'FAILED'
        print_report(output, :bold_error)
      when :interrupt
        log.error 'INTERRUPTED'
        print_report(output, :bold_error)
      else
        log.error '???'
        print_report("Hook returned unknown status `#{status.inspect}` -- ignoring.",
                     :bold_error)
      end
    end
```

Is this the best way to add a `:skip` option to pre_commit hooks, or are there other factors that need to be considered?

**Comment by sds** [comment]
I see where you're going here, but your use of `:skip` in this context goes against the spirit of its purpose in Overcommit.

Skipping a hook is intended to be a temporary thing. In your use case, it makes more sense to completely disable the hook, rather than have the hook detect when it should be run.

Your `before_commit` gem should be able to handle this easily. It can edit the `.overcommit.yml` file to enable the `Brakeman` hook for Rails projects. This is better because it saves you from having to load and run the hook code at all, and allows you to simplify your hook code to not have to check if it should be run.

For now, if you just want to remove the distracting output, you can add `quiet: true` to the hook's configuration so it will print output only if the hook warns/fails.

To summarize:
- The `:skip` status goes against the purpose of skipped hooks in that it should be an intentional decision by the developer when they are about to commit
- You can change `before_commit` to enable the `Brakeman` gem only for Rails projects in your `.overcommit.yml`
- `quiet: true` will silence hook output unless it failed

Hope that helps!

**Comment by reggieb** [comment]
OK. Personally, I think being too prescriptive about how a gem is used is a mistake, but I respect your right to make that decision. We can always use our own fork to add that functionality, though we'll probably just live with the `:pass` functionality.

**Comment by sds** [comment]
> I think being too prescriptive about how a gem is used is a mistake

Any software library is just another tool–I wouldn't give you a hammer and stay quiet when you decide to hit the nail with the wrong end.

The approaches I suggested are quite reasonable, and actually fit in with how the tool is supposed to be used. You shouldn't be running hooks which decide whether they should skip–they should be disabled from the get-go if a repository doesn't require the hook. Ignoring this results in slower hook runs since you have to load extra code which ultimately doesn't get executed. Disabling the hooks you don't use prevents this.

Hope that clarifies the reasoning here.

**Comment by thegranddesign** [comment]
@sds For what it's worth, I also hit on this same issue.  I want to have a common `.overcommit.yml` that I can copy to all of my Ruby projects.  The brakeman integration is the only one that only plays nicely with Rails but not with Rubygems.

I think considering the Brakeman check passed if a Rails project is not found is completely reasonable as that plugin is only valid if run within a Rails project.

Now that I'm thinking about it, I think the source of the issue is with brakeman itself.  It shouldn't be erroring out if a Rails project is not found.

As a compromise, if I can't get them to return a `0` status code (which may be a tall order considering it probably needs to stay an error status for backwards compatibility), would you be willing to consider the hook passable if I could get a status code returned that _only_ meant "There is nothing to do".  So it would basically be like any other linter and technically what brakeman _should_ be doing if it's run outside of a Rails project.

**Comment by thegranddesign** [comment]
@sds, the brakeman maintainer was kind enough to add `4` as the exit code for "No Rails app detected", which you all should be able to catch fairly easily and prevent the Brakeman hook from throwing an error.

**Comment by thegranddesign** [comment]
It should be safe to add to overcommit now since that error code is not currently used, checking for it will simply cause the current behavior of blowing up, however is the user is on the new brakeman version where that error code it used, it will be caught and not blow up.

**Comment by JacobEvelyn** [comment]
We actually go a step beyond @thegranddesign and we have a single `.overcommit.yml` file that's symlinked into each repo. This helps us keep universal code styles (and allows those styles to change and evolve as our team decides) across many different repos.

I came upon this thread after finding https://github.com/presidentbeef/brakeman/issues/869, when one of our non-Rails repos was giving this error after enabling Brakeman.

@sds I'm curious to hear your thoughts on how we might keep the convenience and consistency of our current setup while adding Brakeman support to all of our Rails repos.

**Comment by sds** [comment]
@JacobEvelyn We'd love to add support to Overcommit allowing you to specify gems as hook plugins (specified via your project's Gemfile, so even Git repos would be possible). This would allow you to easy `bundle update <plugin gem list>` to update your configuration for each respective repository. Understandably this could get annoying if you were making frequent changes to your global Overcommit configuration (which my anecdotal experience tells me is not likely).

The main reason we don't want to provide first-class support for loading your hooks from an external (read: not tracked by the Git repo) location is because this goes against the core spirit and value proposition of the tool. Overcommit provides a way for you to enforce a policy that _all_ developers of the given repository must follow. 

If you sidestep this you're allowing developers to define one-off workarounds or hacks that are **not representative of your organizational policies**. You can choose to enforce them only in CI, but there is value in getting early feedback at commit time. We've been open to accepting pull requests adding basic support for this in https://github.com/brigade/overcommit/issues/435#issuecomment-252118374, but would prefer that this be done in a way such that an organization must opt-in to it explicitly.

In your case, your workaround is endorsed by the organization by virtue of you adding `.overcommit.yml` to your `.gitignore` file.

---

Anyway, I got sidetracked a bit, but I wanted to call out your clever use of symlinking as perhaps antithetical to the tool's design. Happy to accept a pull request that adds special handling for exit status 4 to the `Brakeman` hook, which I think solves your particular use case.
